src/alloc.rs

```diff
@@ -47,31 +44,37 @@ impl<T, A: Allocator> Alloc<T, A> {
 impl<T, A: Allocator> RawMem for Alloc<T, A> {
     type Item = T;
     fn allocated(&mut self) -> &mut [Self::Item] {
-        unsafe { std::slice::from_raw_parts_mut(self.ptr.as_ptr(), self.len) }
+        unsafe { slice::from_raw_parts_mut(self.ptr.as_ptr(), self.cap) }
     }
 
     unsafe fn grow(
         &mut self,
-        cap: usize,
+        addition: usize,
         fill: impl FnOnce(&mut [MaybeUninit<Self::Item>]),
     ) -> Result<&mut [Self::Item]> {
-        let req_cap = self.len.checked_add(cap).ok_or(CapacityOverflow)?;
-        let new_layout = Layout::array::<T>(req_cap)?;
+        let cap = self.cap.checked_add(addition).ok_or(CapacityOverflow)?;
+        let new_layout = Layout::array::<T>(cap).map_err(|_| CapacityOverflow)?;
+
         let ptr = if let Some((ptr, old_layout)) = self.current_memory() {
-            self.alloc.grow(ptr, old_layout, new_layout)?
+            self.alloc.grow(ptr, old_layout, new_layout)
         } else {
-            self.alloc.allocate(new_layout)?
-        };
-        let uninit = NonNull::slice_from_raw_parts(ptr.cast::<T>(), req_cap)
-            .get_unchecked_mut(self.len..)
+            self.alloc.allocate(new_layout)
+        }
+        .map_err(|_| AllocError { layout: new_layout, non_exhaustive: () })?
```

Comment:
`AllocError` from `std::alloc::AllocError` does not say anything yet, at least let it store `Layout`